### ISSUE TITLE
[RFC] Compression for inputs (most obviously, UDP input)

### ISSUE THREAD

**Issue by daviesalex** [influxdata/influxdb#4656]
It seems the consensus is that at scale, the UDP input should be used. 

If batching properly, the data from real world setups is extremely compressible. Adding native support to pass in compressed data (with whatever algorithm makes most sense for Go) would substantially reduce network traffic across WAN links and should be able to run efficiently in multiple goroutines.

Once added, the next step would be to modify telegraf to compress data on the way in, (telegraf does not yet support UDP, but will do shortly - see https://github.com/influxdb/telegraf/issues/339)

**Comment by dswarbrick** [comment]
FWIW, we were originally feeding data into InfluxDB using collectd's native UDP protocol, and had a fair bit of trouble with the InfluxDB server not pulling data from the socket fast enough, which resulted in the kernel's udp packet receive buffers eventually filling, overflowing, and then dropping packets. The end result was a lot of gaps in our graphs.

We've had a lot more success with using HTTP to send data, using an inhouse-developed "write_influxdb" collectd output plugin. We consistently achieve about 10:1 compression using GZIP encoding, and due to the inherent properties of TCP, no more dropped packets, no more gaps in our graphs.

Since UDP is stateless, and essentially knows nothing about the packet before, nor the packet after, any compression would have to be constrained to a single packet. GZIP has (by default) a window size of 32 KB, i.e., essentially anything within the previous 32 KB is a candidate for compression. Operating on a small input size, such as a single UDP packet, is not going to yield such good compression ratios.

**Comment by daviesalex** [comment]
So @pauldix told us that we should move, as soon as Telegraf supports it, away from HTTP and onto UDP for large scale writing. I personally have a small preference for UDP; if there are problems with it not working I guess we will have to deal with them in future. In principle, it should be a lot easier to handle a very large number of clients with UDP (far less work for the kernel to do).

To dig into the compression ratios, I took a telegraf instance running on one of our nodes and a tcpdump from it. We have it set to batch 6 sets of metrics and send them every minute (but would prefer to reduce this). Those 6 metrics are 214816 bytes, and if I remove the HTTP headers its 214418 bytes (2826 metrics).

If we removed the batching (which we dont want to do; we are only doing this in a desperate attempt to keep the server up for now) thats about 35kb per "batch" (this nicely fits into a UDP frame with max of 64k), If I take 500 real world metrics from that file into a file (37841 bytes) it reduced by an order of magnitude to 4916 bytes with simple gzip. With xz, its 3984 bytes. This is a huge win even given the small size. I'm sure we could achieve even better compression with some tuning of the algorithm which I have not attempted.

**Comment by dswarbrick** [comment]
@daviesalex I'm curious, what kind of network technology you are using that can transport a 64 KB UDP packet? Most networks are still using MTU of 1500. Even 9K jumbo frame environments are typically only found in data centers, and you're basically screwed as soon as you need to talk to something outside the data center.

xz definitely offers better compression than gzip, but is quite a lot more CPU hungry on the compressing side.

**Comment by daviesalex** [comment]
@dswarbrick, I am purely hypothesizing here because right now we are not using it, but i'd expect the UDP frame to be passed to the kernel in one go and then fragmented by the IP layer, and then the kernel on the Influx server to reassemble it (if any fragment fails to arrive, the whole lot is lost, which isnt a big deal for this use case). In our particular use case, we could turn Jumbo frames on (we have it enabled on switches), but thats largely irrelevant I think.

So from the POV of InfluxDB its sending and receiving 64k packets, which are being chopped down to ~1500 byte Ethernet frames. Of course, in this case it would be sending and receiving compressed 64k, which goes down to about 4000 bytes, which would be sent (and received) as 3 frames.

I could be completely wrong so please do correct me if so!

**Comment by dswarbrick** [comment]
We've had pretty good success with our collectd write_influxdb output plugin filling a 256 KB buffer with data point lines, and gzipping that. This will usually compress down to about 22 KB.

You lose a fair bit of functionality by using UDP, or at least, you have to reinvent a lot of wheels. How about the ability to proxy it? What about crypto? Authentication? These are all inherent in HTTP. Even just determining the maximum size UDP packet that your network will support can be tricky with UDP. TCP has PMTUD, but it's a guessing game with UDP. Whilst the packets _can_ be up to 64 KB, the standard does not actually mandate that all IP stacks _support_ such large UDP packets. On the contrary, RFC 791 (I know, it's prehistoric) only specifies that hosts must be able to receive packets up to 576 octets.

If you're going to rely on the IP stack to reassemble UDP fragments, you're rapidly heading in the direction of TCP anyway. About the only difference is that TCP is going to retransmit lost segments, or may hang if there are network issues between sender and receiver. These problems can be mitigated by using short timeouts however.

IMHO, a UDP protocol would only be interesting if it were binary, i.e. something like ProtoBuf... but... didn't InfluxDB already throw that out, in an earlier version?

**Comment by otoolep** [comment]
FWIW, @sparrc is investigating improving UDP performance since there is still some question about its performance. We think we can service the socket even faster, which may help.

**Comment by daviesalex** [comment]
We moved to UDP yesterday and have found a huge increase in stability. How much of this is because of UDP itself, and how much because of the way UDP is handling. The kernel is happily handling our traffic without any noticeable load and unlike with TCP we have not had to make any effort to tune it to get this performance. I'm not sure how this compares to the setup of @dswarbrick ?

Today, we are injecting 200 megabit of metrics per second from ~4k telegraf clients with a low and stable CPU usage (using about 6 cores). With HTTP we struggled to get over 50 MBit/sec and keep the server up for more than a few hours (we found the process simply became unresponsive) and we were dealing with a system load of well over 30 (we are using 36 core boxes).

Given that we will be sticking with UDP as a input, UDP compression would be a no-brainer for us to reduce network traffic longer term. UDP is significantly simpler for us to scale, so there is just no reason for us to move back to HTTP that I can think of.

Tagging @sparrc to confirm his UDP work on Telegraf works!
